openvino_xai/methods/white_box/torch.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copy & edit from https://github.com/openvinotoolkit/training_extensions/blob/2.1.0/src/otx/algo/explain/explain_algo.py
+"""Algorithms for calculcalating XAI branch for Explainable AI."""
+
+import copy
+from typing import Any, Callable, Dict, Mapping
+
+import numpy as np
+import torch
+
+from openvino_xai.common.utils import SALIENCY_MAP_OUTPUT_NAME
+from openvino_xai.methods.base import IdentityPreprocessFN, MethodBase
+
+
+class TorchMethod(MethodBase[torch.nn.Module, torch.nn.Module]):
+    """
+    Base class for Torch-based methods.
+
+    :param model: Input model.
+    :type model: torch.nn.Module
+    :param preprocess_fn: Preprocessing function, identity function by default
+        (assume input images are already preprocessed by user).
+    :type preprocess_fn: Callable[[np.ndarray], np.ndarray]
+    :parameter target_layer: Target layer (node) name after which the XAI branch will be inserted.
+    :type target_layer: str
+    :param embed_scaling: Whether to scale output or not.
+    :type embed_scaling: bool
+    :param device_name: Device type name.
+    :type device_name: str
+    """
+
+    def __new__(cls, *args, **kwargs):
+        return object.__new__(cls)
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        preprocess_fn: Callable[[np.ndarray], np.ndarray] = IdentityPreprocessFN(),
+        target_layer: str | None = None,
```

Comment:
should `target_layer` be a required parameter as for now?

Reply:
Right, but I'm implementing the layer detection in case for `None`.
Let's keep it as is. :)

Reply:
Sure, but some check for `target_layer is None` is needed anyway somewhere. If layer detection will fail, e.g. it might return `None`, in this case we need to catch it.